include/CL/opencl.hpp

```diff
@@ -6993,52 +6992,47 @@ inline Program linkProgram(
         &error_local);
 
     detail::errHandler(error_local,__COMPILE_PROGRAM_ERR);
-    if (err != NULL) {
+    if (err != nullptr) {
         *err = error_local;
     }
 
     return Program(prog);
 }
 
 inline Program linkProgram(
-    vector<Program> inputPrograms,
-    const char* options = NULL,
-    void (CL_CALLBACK * notifyFptr)(cl_program, void *) = NULL,
-    void* data = NULL,
-    cl_int* err = NULL) 
+    vector<Program>& inputPrograms,
+    const char* options = nullptr,
+    void (CL_CALLBACK * notifyFptr)(cl_program, void *) = nullptr,
+    void* data = nullptr,
+    cl_int* err = nullptr)
 {
     cl_int error_local = CL_SUCCESS;
-
-    vector<cl_program> programs(inputPrograms.size());
-
-    for (unsigned int i = 0; i < inputPrograms.size(); i++) {
-        programs[i] = inputPrograms[i]();
-    }
-    
     Context ctx;
+
     if(inputPrograms.size() > 0) {
         ctx = inputPrograms[0].getInfo<CL_PROGRAM_CONTEXT>(&error_local);
         if(error_local!=CL_SUCCESS) {
             detail::errHandler(error_local, __LINK_PROGRAM_ERR);
         }
     }
+
     cl_program prog = ::clLinkProgram(
         ctx(),
         0,
-        NULL,
+        nullptr,
         options,
-        (cl_uint)inputPrograms.size(),
-        programs.data(),
+        static_cast<cl_uint>(inputPrograms.size()),
+        reinterpret_cast<cl_program *>(inputPrograms.data()),
```

Comment:
OK. Done

Reply:
Shall I remove this change: https://github.com/Oskar-Plaska-Mobica/OpenCL-CLHPP/blob/linkProgram_UT/tests/test_openclhpp.cpp#L3120 ?

Reply:
Having it in the unit test is good, but it requires building and running the tests and any problems would only be found at runtime.  The `static_assert` is even better because the compiler will catch the problem at compile time though, so I'd prefer to use a `static_assert` in this case.

I'd be fine including this in a separate PR if you'd like.  We could add it to your unit test PR for example, or do a dedicated PR just for static asserts (I think we could also use some for `cl::Event` vs. `cl_event`, for example).

Reply:
Yes you are right, static_assert is better than UT.

I can modify my UT PR, so you want me to modify: https://github.com/Oskar-Plaska-Mobica/OpenCL-CLHPP/blob/linkProgram_UT/tests/test_openclhpp.cpp#L3120 and use static_assert instead.
Did I understand your idea correctly ?

Reply:
I think the `static_assert` should go in `opencl.hpp`, not in the unit test.